scriptmodules/emulators/mupen64plus/mupen64plus.sh

```diff
@@ -0,0 +1,265 @@
+                        local dir="${ini_value:2}"
+                        ini_value="${ini_value:1}"
+                        case $dir in
+                            up)
+                                dir="1"
+                                ;;
+                            right)
+                                dir="2"
+                                ;;
+                            down)
+                                dir="4"
+                                ;;
+                            left)
+                                dir="8"
+                                ;;
+                        esac
+                        m64p_hotkey+="H${ini_value:0:1}V${dir}/"
+                        break
+                    else
+                        m64p_hotkey+="B${ini_value//\"/}/"
+                        break
+                    fi
+                ;;
+            esac
+        fi
+    done
+
+    # search hotkey
+    for input_type in "_btn" "_axis"; do 
+        # add hotkey and append enable hotkey button
+        # return if hotkey exists
+        iniGet "${key}${input_type}"
+        ini_value="${ini_value// /}"
+        if [[ "$ini_value" != "" ]]; then
+            case "$input_type" in
+                _axis)
+                    ini_value="${ini_value//\"/}"
+                    m64p_hotkey+="A${ini_value:1}${ini_value:0:1}"
+                    echo "$m64p_hotkey"
+                    return
+                    ;;
+                _btn)
+                    # if ini_value contains "h" it should be a hat device
+                    if [[ "$ini_value" == *h* ]]; then
+                        ini_value="${ini_value//\"/}"
+                        local dir="${ini_value:2}"
+                        ini_value="${ini_value:1}"
+                        case $dir in
+                            up)
+                                dir="1"
+                                ;;
+                            right)
+                                dir="2"
+                                ;;
+                            down)
+                                dir="4"
+                                ;;
+                            left)
+                                dir="8"
+                                ;;
+                        esac
+                        m64p_hotkey+="H${ini_value:0:1}V${dir}"
+                        echo "$m64p_hotkey"
+                        return
+                    else
+                        m64p_hotkey+="B${ini_value//\"/}"
+                        echo "$m64p_hotkey"
+                        return
+                    fi 
+                    ;;
+            esac
+        fi
+    done
+    return
+}
+
+function remap() {
+    local devices
+    local device_numbers
+    local device_counter=0
+    local device_number
+    
+    # get lists of all present js device numbers and device names
+    # get device count
+    for device_number in {0..10}; do
```

Comment:
instead of hardcoding and looping through 0-10 you could loop through the actual existing devices (using ls or so) - there would be no need to check if the device existed then (if you need the id # you can easily extract that if needed)